kuma/javascript/src/document-provider.jsx

```diff
@@ -0,0 +1,117 @@
+// @flow
+import * as React from 'react';
+import { useState, useEffect } from 'react';
+
+export type DocumentData = {
+    locale: string,
+    slug: string,
+    id: number,
+    title: string,
+    summary: string,
+    absoluteURL: string,
+    redirectURL: string,
+    editURL: string,
+    bodyHTML: string,
+    quickLinksHTML: string,
+    tocHTML: string,
+    parents: Array<{ url: string, title: string }>,
+    translations: Array<{ locale: string, url: string, title: string }>
+};
+
+const context = React.createContext<DocumentData | null>(null);
+
+type DocumentProviderProps = {
+    children: React.Node,
+    initialDocumentData: DocumentData
+};
+
+export default function DocumentProvider(
+    props: DocumentProviderProps
+): React.Node {
+    const [documentData, setDocumentData] = useState(props.initialDocumentData);
+
+    if (!document.body) {
+        throw new Error('DocumentProvider rendered before a body exists.');
+    }
+    const body = document.body;
+
+    // A one-time effect that runs only on mount, to set up
+    // an event handler for client-side navigation
+    useEffect(() => {
```

Comment:
This is the function that sets up client-side navigation. The click event listener is what does most of the work.